R Notebook for R Review - ANSWER KEY
QBIO Public Data Analysis Student Group
3/12/2021

Here is a review R Notebook. This is designed to be useful for the small group project. 
```{r}
#if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
#if(!requireNamespace("devtools")) BiocManager::install(c("devtools"))
#if(!requireNamespace("robustbase"))BiocManager::install(c("robustbase"))
#library(devtools)
#library(robustbase)

#devtools::install_github("BioinformaticsFMRP/TCGAbiolinks")

if(!requireNamespace("SummarizedExperiment"))BiocManager::install(c("SummarizedExperiment"))
if(!requireNamespace("HDF5Array"))BiocManager::install("HDF5Array")

library(TCGAbiolinks)
library(SummarizedExperiment)
library(HDF5Array)
```

Use these barcodes if using on your local machine as an example for smaller files downloaded
```{r}
barcodes_rnaseq <- c("TCGA-BH-A0DG-01A-21R-A12P-07","TCGA-A2-A0YF-01A-21R-A109-07",
          "TCGA-AN-A04A-01A-21R-A034-07","TCGA-AR-A1AP-01A-11R-A12P-07",
           "TCGA-A2-A0T3-01A-21R-A115-07", "TCGA-E2-A154-01A-11R-A115-07" )
barcodes_clinic <- c("TCGA-BH-A0DG","TCGA-A2-A0YF","TCGA-AN-A04A","TCGA-AR-A1AP", "TCGA-A2-A0T3",
                      "TCGA-E2-A154", "TCGA-AO-A12F", "TCGA-A2-A0YM", "TCGA-BH-A0E0", "TCGA-AN-A0FL")
```

PART 1: DATAFRAMES

Load in the subtype data
1. Use head() to look at the subtypes dataframe
  a. What columns are in subtypes?
  b. Which columns do you personally find interesting and why?
```{r}
subtypes <- TCGAquery_subtype(tumor = "BRCA")
head(subtypes)
colnames(subtypes)
```
2. Use str() and dim() to look at subtypes
  a. How is str() different than head?
      head() gives the first 5 entries of the table. str() gives an overall structure of the dataframe
  b. How many rows and columns are in subtypes?
      1087 rows and 24 columns
  c. How many patients are in subtypes?
      1087 patients
```{r}
str( subtypes )
dim( subtypes )
```

3. Standard Indexing
  a. Use the format, dataframe[row, col] to find the element at row 4 and col 3.
  b. Removing the row or col in [row,col] allows you to access an entire row or column. Access all columns of the 7th row. 
  c. 1:10 is a special R way to create a list of from 1 to 10. Type 1:10 in the console if this is unclear. 
     Access all columns of the first 10 rows. Note this is the first 10 patients.
  d. Is the format subtypes[row,col] useful for this dataframe? Why or why not?
          Not especially. Row number and column number do not have a clear meaning in this context. Instead using $ or boolean indexing (next) are more useful. Note: There are dataframes that have row names. In this case it may be helpful to use the [row,col] formatting with [rowname, colname]. Try subtypes[, "patients"] as an example of using the name of a column
  e. Don't forget about the $. Create a variable called BRCA_PAM50 and assign the BRCA_Subtype_PAM50 column to it. 
```{r}
#a
subtypes[4,3]

#b
subtypes[7,]

#c
1:10
subtypes[1:10,]

#e
BRCA_PAM50 <- subtypes$BRCA_Subtype_PAM50

```

4. Boolean Indexing
Boolean Indexing is a unique and helpful way to index into a dataframe.
It utilizes a "boolean mask". A boolean is a TRUE or FALSE element. A boolean mask is a list of TRUE's and FALSE's.
Follow the following steps to find patients less than 40yrs old AND are in the LumA subtype
  1. Describe the contents of luma_mask. Note that "==" means "equal to"
        luma_mask is a list of TRUE's and FALSE's. It is TRUE when the patient is Luminal A subtype.
        
  2. Write an equivalent line of code that creates the variable luma_mask_check that is equivalent to luma_mask,
     but uses the BRCA_PAM50 variable above.
     
  3. Computers read TRUE as "1" and FALSE as "0". A cool trick is to sum the elements of the mask to see how many TRUE elements exist.
     Using sum() and as.numeric(luma_mask), how many patients are Luminal A subtype?
     562 patients are Luminal A
         
  4. Use information from ^^ check that luma_mask and luma_mask_check have the same number of TRUE elements.
  
  5. Create a boolean mask called young_mask that is TRUE when the patient is less than (<) 40 yrs old. How many patients are under 40?
     How many patients are over 40 in subtypes?
     76 patients are under 40. 1011 patients are over 40 years old. 
     
  6. Create a boolean mask called young_and_luma_mask that is TRUE when the patient in LumA and young. Use the "&" symbol.
  
  7. You can use the boolean mask to index into the subtypes dataframe. Remember part 3. Standard Indexing. 
     Create a new dataframe called young_and_luma_dataframe and assign it to all columns of the patients that are young and luminal a
     using your young_and_luma_mask. Hint: This is the same format as the 3c using the 1:10. 
     
  8. What if we just want which patients are young and luminal a? You can also index into a specific column. 
     Using $ and subtypes[young_and_luma_mask], create a variable young_and_luma_patients with the list of young and luminal a patients.
```{r}
#1
luma_mask <- subtypes$BRCA_Subtype_PAM50 == "LumA"

#2
luma_mask_check <- BRCA_PAM50 == "LumA"

#3
sum(as.numeric(luma_mask))

#4
sum(as.numeric(luma_mask)) == sum(as.numeric(luma_mask_check))

#5
young_mask <- subtypes$age_at_initial_pathologic_diagnosis < 40
sum(as.numeric(young_mask))
dim(subtypes)[1] - sum(as.numeric(young_mask))

#6
young_and_luma_mask <- luma_mask & young_mask

#7
young_and_luma_dataframe <- subtypes[ young_and_luma_mask, ]

#8
young_and_luma_patients <- subtypes$patient[ young_and_luma_mask ]
```

4.5. Use masks to summarize data.
  a. What percentage of the young (<40) patients in the subtypes data frame are Luminal A?
  b. What percentage of the old (>=60) patients in the subtypes data frame are Luminal A?
```{r}
#mutliple ways to do this
#Doing something multiple ways is a good sanity check to make sure your answer is correct and your calculation makes sense
young_patients_dataframe <- subtypes[ young_mask, ]
num_luma_young <- sum( as.numeric(young_patients_dataframe$BRCA_Subtype_PAM50 == "LumA") )
num_young <- sum(as.numeric(young_mask))
percent_of_young_luma <- num_luma_young / num_young * 100
percent_of_young_luma

#another way
as.numeric(sum(young_and_luma_mask)) / sum(as.numeric(young_mask)) * 100

#another way
as.numeric(sum(young_patients_dataframe$BRCA_Subtype_PAM50 == "LumA")) / dim(young_patients_dataframe)[1] * 100

#b is the same but with old patients >= 60. 
```


PART 2: RNASeq data

Download and/or access data
```{r}
query <- GDCquery(project = "TCGA-BRCA",
                   data.category = "Transcriptome Profiling",
                   data.type = "Gene Expression Quantification",
                   workflow.type = "HTSeq - Counts",
                   barcode = barcodes_rnaseq)
GDCdownload(query) #only need this line of code ONCE to download the data
sum_exp <- GDCprepare(query)
```

1. SummarizedExperiment 
   a. What does sum_exp include?
      sum_exp is a SummarizedExperiment with colData, rowData, and assays
   b. What does colData(sum_exp) include?
      colData( sum_exp ) has information about the patients. The patients are the rows and columns are different information about the patients. 
   c. What does rowData(sum_exp) include?
      rowData(sum_exp) has information about the genes. Genes are the rows, and columns are different gene names.
   d. What does assays(sum_exp) include?
      assays(sum_exp) can include different experimental data. Here it only contains one type of data (HTSeq - Counts)
   e. Describe assays(sum_exp)$"HTSeq - Counts"
      This has the ensembl gene ids as the rows and patients as the columns. THe table is filled with the gene expression counts.
```{r}
sum_exp
head( colData(sum_exp) )
head( rowData(sum_exp) )
head( assays(sum_exp)$"HTSeq - Counts" )
```

2. ifelse
Remember the ifelse structure:
ifelse(condition, action when condition true, action when condition false)
   a.Create a new column in colData(sum_exp) called age_category with the following age categories:
     < 40 is "young", 40-59 is "mid", 60+ is "old".
     Use colData(sum_exp)
```{r}
colData(sum_exp)$age_category <- ifelse(colData(sum_exp)$paper_age_at_initial_pathologic_diagnosis < 40, "Young", ifelse(colData(sum_exp)$paper_age_at_initial_pathologic_diagnosis >= 60, "Old", "Mid"))
```

3. Applying boolean masks to HTSeq
   a. Create variable young_htseq_mask using colData(sum_exp)$age_category
   b. Create variable htseq_data that consists of: assays(sum_exp)$"HTSeq - Counts"
   c. Create variable young_htseq_data that consists of all genes (rows) of only the young patients (cols). Run str(young_htseq_data).
   d. Use dim(young_htseq_data)[2] == sum(as.numeric()) to check that your code is correct.
   e. Why does (d) verify if your code is correct?
      This verifies if the code is correct because dim() returns a list of 2 elements, [# of rows, #of columns]. dim()[2] is the number of columns which is the number of patients. By using == we check that the number of columns (number of patients) in our young dataframe is the same as the number of young patients identified by the young_htseq_mask
```{r}
#a
young_htseq_mask <- colData(sum_exp)$age_category == "Young"

#b
htseq_data <- assays(sum_exp)$"HTSeq - Counts"

#c
young_htseq_data <- htseq_data[, young_htseq_mask]

#d
dim(young_htseq_data)[2] == sum(as.numeric(young_htseq_mask))
```

4. Note about barcodes
Notice that in colnames(colData(sum_exp)) there are two columns that discuss barcodes: 
                                                    colData(sum_exp)$barcode and colData(sum_exp)$bcr_patient_barcode
Notice the barcode is longer with extra numbers and letters. This is because it refers the the tumor SAMPLE.
The bcr_patient_barcode refers to the PATIENT. You can imagine that one PATIENT can have multiple tumor SAMPLES.
The bcr_patient_barcode is the same in the clinical data, and the subtypes dataframe above. 
In your project, you can use this barcode to interface between breast cancer subtype (an important factor in breast cancer) and the RNASeq counts.
Alternatively, you can use the colData(sum_exp)$paper_BRCA_Subtype_PAM50 column in sum_exp
```{r}
head( colData(sum_exp)$barcode )
head( colData(sum_exp)$bcr_patient_barcode )
head( colData(sum_exp)$paper_BRCA_Subtype_PAM50 )
```


5. Note about saving the SummarizedExperiment
As you perform your analysis, you can reload in your TCGA HTSeq data through the query function as we have been doing. 
Alternatively, you can do the query once, and save the SummarizedExperiment as an HDF5Array
   1. Use getwd() to print out your current PATH. This is where you are currently located in your system. 
   2. Whenever you are saving or loading in files, you can use setwd() to specify the PATH you wish to reference. 
   3. Use saveHDF5SummarizedExperiment() to save your SummarizedExperiment in desired directory (dir). Remember, this is RELATIVE to your current location or PATH that you just specified in setwd(), or looked at in getwd(). 
   
Source: https://bioconductor.org/packages/release/bioc/manuals/HDF5Array/man/HDF5Array.pdf

```{r}
getwd()
setwd("C:/Users/Kate/qbio_data_analysis/")

#the dir creates a directory with this name. 
saveHDF5SummarizedExperiment(sum_exp, dir="htseq_h5_sumexp", prefix="", replace=FALSE,chunkdim=NULL, level=NULL, as.sparse=NA,verbose=NA)

#To load in your hdf5 summarizedexperiment
sum_exp_loaded <- loadHDF5SummarizedExperiment(dir="htseq_h5_sumexp", prefix="")
str(sum_exp_loaded)
```



